### ISSUE TITLE
Paramiko logs passphrase

### ISSUE THREAD

**Issue by tgrgin** [paramiko/paramiko#957]
With log-level DEBUG, Paramiko logs the passphrase for PK:
2017-05-09 13:20:47,723: DEBUG [_log;transport.py:1572]: Trying discovered key with passphrase MY_REAL_PASSPHRASE!
2017-05-09 13:20:47,723: DEBUG [_log;transport.py:1572]: Trying discovered key 070xxx14a in
C:\Users\user/.ssh/id_rsa
2017-05-09 13:20:47,723: DEBUG [_log;transport.py:1572]: userauth is OK

I would like for Paramiko not to log any sensitive information.
Environment: Python 2.7.13, Paramiko 2.1.2, Windows.

**Comment by ploxiln** [comment]
Can you provide any more details, like what paramiko-using program you were running?

In paramiko 2.1.2 I can find the word "discovered" in only one place in the codebase:

```
            for pkey_class, filename in keyfiles:
                try:
                    key = pkey_class.from_private_key_file(filename, password)
                    self._log(DEBUG, 'Trying discovered key %s in %s' % (hexlify(key.get_fingerprint()), filename))
```

I can find the word "passphrase" in just a few places, and outside of a test and a demo (demo_keygen.py) I can't find it anywhere that it could appear in any string used by the program.

When I use fabric-1.13.2 and paramiko-2.1.2 and enable paramiko debug logging, and use a private key file with a passphrase, this is what I get:

```
...
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Adding ssh-rsa host key for testdeploy05.ec2.st-av.net: 960b1437305b03aa89ee7aee505c8cb3
[testdeploy05.ec2.st-av.net] Passphrase for private key: 
DEBUG:paramiko.transport:starting thread (client mode): 0x18aab50L
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.1.2
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.1
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.2p2)
DEBUG:paramiko.transport:kex algos:[u'curve25519-sha256@libssh.org', u'ecdh-sha2-nistp256', ...
DEBUG:paramiko.transport:Kex agreed: diffie-hellman-group14-sha1
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha2-256
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexGroup14 specified hash_algo <built-in function openssl_sha1>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Trying discovered key 15f03d84c247df57e9e8998a6f0a9edb in /Users/pierce/.ssh/id_rsa
DEBUG:paramiko.transport:userauth is OK
...
```

I typed my passphrase in when it printed the prompt "[testdeploy05.ec2.st-av.net] Passphrase for private key: " but the passphrase itself did not appear. Maybe this is a limitation of integration with the windows command shell?

**Comment by tgrgin** [comment]
Ploxiln, thanks for your quick response. Thing is, you are entering passphrase when asked while I have JS UI where user sets connection up. Thus, I call:
c = paramiko.SSHClient()
...
c.connect(hostname=self.host, username=self.user, password=self.pwd, pkey=newKey)
or
c.connect(hostname=self.host, username=self.user, password=self.pwd)
as per http://docs.paramiko.org/en/2.1/api/client.html:
connect(hostname, port=22, username=None, password=None, pkey=None, key_filename=None, timeout=None, allow_agent=True, look_for_keys=True, compress=False, sock=None, gss_auth=False, gss_kex=False, gss_deleg_creds=True, gss_host=None, banner_timeout=None)
...
Parameters:	
...
    password (str) – a password to use for authentication or for unlocking a private key

So, to reproduce this problem one has to call client.connect and pass the passphrase.

**Comment by ploxiln** [comment]
If I just specify `password` (which can decrypt `~/.ssh/id_rsa`) to `SSHClient.connect()`:

```
DEBUG:paramiko.transport:Kex agreed: ecdh-sha2-nistp256
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha2-256
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Trying discovered key 8dd08b296963b43ca24ec49fab018da7 in /Users/pierce/.ssh/id_rsa
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (publickey) successful!
```

The other possible ways of specifying a key don't log with the word "discovered". But anyway, here's what it logs if I load the private key beforehand (at which point it is decrypted) and then pass it for pkey:

```
DEBUG:paramiko.transport:Kex agreed: ecdh-sha2-nistp256
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha2-256
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Trying SSH key 8dd08b296963b43ca24ec49fab018da7
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (publickey) successful!
```

The obvious place in the code of paramiko-2.1.2 does not obviously log any passphrase. I did search (grep) the entire codebase to look for places that might log "discovered" or "passphrase" and didn't find anything suspicious.

So this must be from code outside paramiko. Maybe some code outside paramiko calls `SSHClient._log()` or `Transport._log()` with the passphrase?

**Comment by bitprophet** [comment]
Yea, I've never seen that particular log output before, nor can I find an example of it in Git history (though that would've been unlikely given the declared version.)

Going to close as "something else must be interfering" but definitely curious how this is sneaking in - would love more info about exactly how @tgrgin is obtaining, installing & calling Paramiko in this case, and what other Python code is involved. Feels like client code is calling Paramiko's logger (as per @ploxiln) or perhaps something else installed in the same environment is monkeypatching us?

**Comment by tgrgin** [comment]
Thank you guys, will investigate further.

**Comment by tgrgin** [comment]
Could it be it is a problem in Paramiko:
```
PS:573 [FRIGG14]> Get-ChildItem -Path . -Include *.py -Recurse | 
    Select-String -Pattern "Trying discovered key with passphrase"

Python\Lib\site-packages\paramiko\client.py:600: 
    self._log(DEBUG, 'Trying discovered key with passphrase %s' % (password))
```

```
F:\MCCgit\WLTEST2\share\mcc\Python>python.exe
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import paramiko
>>> print paramiko.__version__
2.1.2
```

**Comment by tgrgin** [comment]
bitprophet, everything is clean. Python from proper place, Paramiko via pip, absolutely no changes to Paramiko code.
As per calling, already posted that:
```
c = paramiko.SSHClient()
...
c.connect(hostname=self.host, username=self.user, password=self.pwd, pkey=newKey)
or
c.connect(hostname=self.host, username=self.user, password=self.pwd)
```

**Comment by radssh** [comment]
Both the wheel install and tarball (verified MD5) from PyPI of paramiko 2.1.2 have the logging line in the source that matches the GitHub source:

```self._log(DEBUG, 'Trying discovered key %s in %s' % (hexlify(key.get_fingerprint()), filename))```
 and the client.py file is (as expected) an exact match with https://raw.githubusercontent.com/paramiko/paramiko/master/paramiko/client.py 

I'd be curious to see what the whole of the diff listing shows between your client.py and the current (2.1.2) source.

**Comment by tgrgin** [comment]
There are several differences more (1 IF several log(DEBUG)) but no time to go into details atm. Probably some other files are changed too so I'll just reinstall Paramiko. In any case, under terms of LGPL, I am not allowed to make changes to your code without asking permission from my Legal and I do not want to go there.

Thanks a lot guys, this was a real concern for me!

**Comment by ploxiln** [comment]
If I was you, I'd be more concerned than ever: Where did that line of code come from?!

You said "Paramiko via pip, absolutely no changes to Paramiko code". That line has _never_ been in any paramiko release! Do you have a pip cache or mirror for your company? Did one of your co-workers put a customized version with more debug logging in the company pip mirror? Or was it "back-doored" by someone trying to infiltrate your systems?

**Comment by tgrgin** [comment]
Ploxiln, I was concerned about the same, just do not have time atm to check on everything. This is a part of small project I inherited. That said, I never did any changes but it is well possible someone in past 7-8 years did and when I formed new repositories couple of days ago I must have had merged some debug revision. This is the best I can come up with :-/
Now I will remove Python entirely from the GiT tree to start with.

Sorry for the disturbance and thank you all very much.
